### PR TITLE
Options documentation and refactoring

### LINKED ISSUE #43: Document and update adflow options
## Type of issue
- Documentation update

## Description
We need to put some effort into documenting all adflow options.
Current list https://mdolab-adflow.readthedocs-hosted.com/en/latest/options.html list is outdated and is missing many options.

## Preferred solution
This documentation should preferably be generated automatically. All options exist in `pyADflow.py` and default values etc. should be set from what is found there. The issue is with the detailed descriptions on what each parameter does. Either we can include a detailed description in `pyADflow.py`, or we can write them elsewhere and combine with an automated process.

### PR DESCRIPTION
**Pull request by sseraj**
## Purpose
This PR adds documentation for all ADflow options. (Closes #43)

The options table is generated automatically from the default options dictionary in pyADflow and the descriptions in options.yaml. Doing this required modifying pyADflow to use ``getOption`` and ``setOption`` from BaseSolver in baseclasses.

There are two changes in this PR that may break backward compatibility with ADflow run scripts:
1) String-type option values are now case sensitive and some of the acceptable string values have changed. This will most commonly affect `equationType`. For example, you must use `"RANS"` instead of `"rans"`.
2) ``autosolveretry``, ``autoadjointretry``, and ``nkcfl0`` are now deprecated. These options have had no function for years but may still be in old scripts.

The full list of affected string-type options is:
```
solutionPrecision
gridPrecision
solutionPrecisionSurface
gridPrecisionSurface
discretization
coarseDiscretization
limiter
smoother
equationType
equationMode
flowType
turbulenceModel
turbulenceOrder
turbulenceProduction
eulerWallTreatment
viscWallTreatment
resAveraging
oversetUpdateMode
timeIntegrationScheme
NKLS
adjointSolver
preconditionerSide
matrixOrdering
globalPreconditioner
localPreconditioner
```

These changes will break the pyAeroStructure tests. @eirikurj will fix the options either for the current tests or with the [testflo update](https://github.com/mdolab/pyaerostructure/pull/32). I will submit a PR to fix the tutorial run scripts after this is merged.

Once this and #41 are merged, we will make a new release and make the backward incompatible changes clear in the release notes.

The formatting of options table itself is not the best, but we can address this in the future. See [here](https://github.com/mdolab/sphinx_mdolab_theme/issues/3) for a possible solution.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [x] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
See the [BaseSolver tests](https://github.com/mdolab/baseclasses/blob/master/tests/test_BaseSolver.py) for how ``getOption`` and ``setOption`` are tested.

## Checklist
_Put an `x` in the boxes that apply._

- [ ] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Comment by nwu63** [comment]
Do we want to address #43 before incrementing the minor version/making a release?

**Comment by sseraj** [comment]
I think #43 has been addressed? Is there something missing from this PR?

**Comment by nwu63** [comment]
I guess I'm referring more to #41, whether we want to wait until that's resolved before releasing a new minor version.

**Comment by sseraj** [comment]
I'm leaning towards separate releases, but it really depends on quickly we can merge #41 (or a new PR) after this one. What do you think @joanibal?

**Comment by joanibal** [comment]
I suggest moving forward with this PR. There hasn't been much activity on #41 for a while. 

I'm not sure what the next steps are for #41, maybe we could talk about it at a maintenance meeting?

File: adflow/pyADflow.py
Comment (joanibal) [inline adflow/pyADflow.py:5105]: >"ignore_options" are NOT actually', ignored

oof, glad we changed that variable name

**Review by joanibal — APPROVED** [review]

File: adflow/pyADflow.py
Comment (nwu63) [inline adflow/pyADflow.py:4555]: Are you planning to update these names?

**Comment by nwu63** [comment]
So I just want to highlight again the fact that this PR is proposing to make all option values that are strings to be lower case for consistency. I'm still not 100% convinced since it will reduce some readability, but I also don't feel strongly either way. But everyone should understand the consequences and voice your opinions accordingly.

Alternatively, with a bit more work we can implement something in baseclasses (say a set of option names) where the values should be case insensitive. Then for get/set options they can be cast to lower and made consistent, but the user would be free to input with whatever case they like. I'm not necessarily advocating for this approach, but if we want to preserve the freedom then we definitely want some sort of systematic approach like this rather than an ad-hoc `str.lower()` in random places.

File: adflow/pyADflow.py
Comment (sseraj) [reply adflow/pyADflow.py:4555]: I decided not to for now to keep the backward incompatible changes to a minimum. I could do it though if you think it's worth changing.

Comment (nwu63) [reply adflow/pyADflow.py:4555]: I personally think if we are to make backwards incompatible changes we might as well do them at once. That's why I also asked about #41, I think it's more work to announce/release changes multiple times rather than all at once. Up to you guys though since I'm not even a maintainer for adflow.

Comment (joanibal) [reply adflow/pyADflow.py:4555]: Good point about "squashing" the backwards incompatible changes into one commit. 

I was still in the mindset of 1 change per PR. 
Is there a way to still do one PR per feature and only make one update on the master branch?
Could we collect some backwards incompatible changes in a separate branch?

**Comment by joanibal** [comment]
I'm curious about the requirement for lowercase options too. 
I assumed it had something to do with the automatic documentation process.

File: adflow/pyADflow.py
Comment (nwu63) [reply adflow/pyADflow.py:4555]: I was not thinking about one commit vs. multiple on the master branch, IMO it's fine to have multiple backwards incompatible changes on the master branch. It's more a matter of when to do the releases since my mindset is that that's when changes are formalized, listed and communicated to end users. It's generally not a good idea to just pull down the latest changes on master, and checking out a tagged release should help remind users to read the release notes.

**Comment by nwu63** [comment]
> I'm curious about the requirement for lowercase options too.
> I assumed it had something to do with the automatic documentation process.

Nope it's not due to that, we just wanted to formalize a list of strings which are accepted as possible choices. I do think it's more explicit and less error-prone to accept only one capitalization rather than all possible capitalizations, but whether that's camelCase or lower case is a matter of readability vs. ambiguity.

So I guess there are three options here
1. force exact string matching, lower case only
2. force exact string matching, "reasonable" capitalization is used
3. arbitrary capitalization is accepted

See with a GUI we wouldn't need to have any of this debate :P anyways my ranked choices are 213 for this.

**Comment by joanibal** [comment]
Oh I see

Hmm, my preference is 132, since other repos have already been updated with option 1. At the end of the day this isn't that big a deal for me.

File: adflow/pyADflow.py
Comment (joanibal) [reply adflow/pyADflow.py:4555]: How would you all feel about making these changes, but instead of causing an error, incorrect capitalization leads to a deprecation warning. 

Then we can Merge this PR but avoid incrementing the major version number.

**Comment by nwu63** [comment]
I don't feel super strongly about this either although if we want to settle on a satisfactory choice now is the time to do it. I'd rather change everything now than revisit this in a year's time.

**Comment by sseraj** [comment]
My preference is also 132, but mainly because I think all lowercase is easiest to remember. I have no problem changing all the scripts again if we choose option 2. I will also gladly work on option 3. I agree that the important thing is that we make an intentional choice now rather than flip-flopping in the future. Let's see what @eirikurj and @anilyil think.

**Comment by eirikurj** [comment]
I do realize that from a coding standpoint that its easier to set and expect everything to be lowercase  in order to avoid having to remember how the option is exactly specified.
Regardless of the option we choose, I also think its not a big ask on the user to resort to the options documentation to check how the options is defined and should be used.
Overall, I dont feel that strongly about this, but my preference is to favor readability, so I would prefer an overall camelCase strategy for both the option name and value.
However, using a strict camelCase rule is probably not going to be possible since there will be some special cases such as `Euler`, and acronyms should probably be capitalized, such as `RANS` or `GMRES`.
To avoid these issues and maintain backwards compatibility we should favor 3.
However, if we want to enforce specific case for names and values and are set on breaking backwards compatibility, I think we should go for 2 (with some exceptions) due to readability.
To close this discussion I think we should meet and discuss pros and cons of each strategy.

File: adflow/pyADflow.py
Comment (nwu63) [inline adflow/pyADflow.py:4492]: I thought these options take in either None (default), str, or list of string? Do we need to add `list` to the possible types?

Comment (sseraj) [reply adflow/pyADflow.py:4492]: Maybe @anilyil can clarify this, but for what I gather:
Options with `family` only accept a string or None.
Options with `families` only accept a list or None.

File: adflow/pyADflow.py
Comment (nwu63) [inline adflow/pyADflow.py:4569]: this could be replaced by `(types.FunctionType, types.BuiltinFunctionType)` I think. Not super important but we are expecting a callable here right?

File: adflow/pyADflow.py
Comment (nwu63) [reply adflow/pyADflow.py:4492]: Yes I think you are right, options that behave like I described (such as `restartFile`) has been updated already.

File: adflow/pyADflow.py
Comment (sseraj) [reply adflow/pyADflow.py:4569]: Good call. I will update this. I'm inclined to not include `types.BuiltinFunctionType` though because it really should not be a built-in function.

Comment (nwu63) [reply adflow/pyADflow.py:4569]: Sure, I wasn't sure what types of functions gets passed in.

File: adflow/pyADflow.py
Comment (nwu63) [inline adflow/pyADflow.py:4517]: Since we are already breaking some backwards compatibility, do we want to "fully" update the options? e.g. `van Albada`, `no limiter` and `k-omega`?

Comment (nwu63) [reply adflow/pyADflow.py:4517]: I'm asking since I see a lot of options already have spaces, like `Runge Kutta` instead of `rungeKutta`.

Comment (sseraj) [reply adflow/pyADflow.py:4517]: I don't see why not at this point. Most of the string options do use spaces so I think it would be nice to make it consistent. What about dashes? We don't use those anywhere. Is it better to stick with spaces?

Comment (nwu63) [reply adflow/pyADflow.py:4517]: I think we should use dashes if that's what done in the literature, e.g. `Wilcox k-omega`, but if it's just two words then we should split with spaces. Overall, I think it's fine to keep option names as single words with camelCase, but the values could be expanded to have spaces (or dashes if appropriate). What do you guys think @eirikurj @anily @joanibal

Comment (sseraj) [reply adflow/pyADflow.py]: I added them for now, but it's easy enough to change back.

**Comment by joanibal** [comment]
Is this ready to be reviewed again?

**Comment by sseraj** [comment]
Yes, it is!

**Review by joanibal — APPROVED** [review]

**Comment by anilyil** [comment]
Thanks for this work @sseraj! Just to be clear; the users will need to update the options that have string values if they were modified in this PR correct? How will we communicate this change with users? In release notes? 

Besides that, the PR looks good to me.

**Comment by sseraj** [comment]
Yes, the users will have to update the string values. To communicate this, I am planning on listing the names of the affected string options in the release notes and linking to the options table for the new values. I can also add the list to the PR description.

File: adflow/pyADflow.py
Comment (nwu63) [inline adflow/pyADflow.py:1892]: Is this still true with ANK?

File: adflow/pyADflow.py
Comment (nwu63) [inline adflow/pyADflow.py:4520]: time-spectral? Not a big deal either way

File: adflow/pyADflow.py
Comment (nwu63) [inline adflow/pyADflow.py:4621]: similarly, non-monotone

File: doc/tutorial.rst
Comment (nwu63) [inline doc/tutorial.rst:49]: These still need to be updated right?

File: tests/reg_tests/reg_default_options.py
Comment (nwu63) [inline tests/reg_tests/reg_default_options.py:33]: These options dictionary should be instantiated using `caseInsensitiveSet` to avoid duplicate entries (this was a bug I ran into a while ago which motivated this change)

**Review by nwu63 — COMMENTED** [review]
A few minor comments

File: adflow/__init__.py
Comment (sseraj) [reply adflow/__init__.py:1]: I agree we should wait for the other PR before making a new release. One of the two PRs will have the version bump though so I was going to keep this here. I can change this back if you prefer.

File: adflow/pyADflow.py
Comment (sseraj) [reply adflow/pyADflow.py:4520]: No hyphen actually seems more common in the literature

**Comment by anilyil** [comment]
> Yes, the users will have to update the string values. To communicate this, I am planning on listing the names of the affected string options in the release notes and linking to the options table for the new values. I can also add the list to the PR description.

Sounds good to me. @nwu63 do you request changes? I can approve this PR when things are settled.

File: adflow/__init__.py
Comment (nwu63) [reply adflow/__init__.py:1]: I think only the commit directly preceding the tag should contain the updated version. That way, if you check the version in Python you should be confident it is _at most_ as old as the tag, i.e. whatever change from the release notes have all been incorporated into the software.

File: adflow/pyADflow.py
Comment (sseraj) [reply adflow/pyADflow.py:1892]: I think so. RK is still the best way to increase the residual after restart.

File: adflow/pyADflow.py
Comment (sseraj) [reply adflow/pyADflow.py:4621]: I was actually thinking of making this one word based on how prevalent 'nonmonotone line search' is compared to 'non-monotone line search'.

Comment (nwu63) [reply adflow/pyADflow.py:4621]: Haha I find `nonmonotone` incredibly illegible, but that's still better than `non monotone` IMO.

File: adflow/pyADflow.py
Comment (eirikurj) [reply adflow/pyADflow.py]: I agree. For string values we should make them more readable by separating using spaces when appropriate.

File: adflow/pyADflow.py
Comment (anilyil) [reply adflow/pyADflow.py:1892]: `'infchangecorrection':True` is also a great option (better than RK reset imo). I think we are planning to make that enabled by default in #41

File: tests/reg_tests/reg_default_options.py
Comment (sseraj) [reply tests/reg_tests/reg_default_options.py:33]: I tried this and it breaks things elsewhere in the tests. I think we should leave this for a later PR.

Comment (nwu63) [reply tests/reg_tests/reg_default_options.py:33]: I'll open an issue for this.

**Review by nwu63 — APPROVED** [review]

File: adflow/pyADflow.py
Comment (anilyil) [inline adflow/pyADflow.py:4672]: Should we also update these options? It should be `resRho` etc. with the current convention right? I think these are more commonly used compared to other affected options, so I am also fine with not changing these.

Comment (nwu63) [reply adflow/pyADflow.py:4672]: I suggest that we keep this for now. In the future, I plan to change the way these are set up in baseclasses. Like the list of choices, there should be a _set_ of all possible variables to monitor, and the user can specify a list (does the order matter? maybe for print out?) and we check in base classes if that is a subset of the total allowable set of monitor variables.

**Review by anilyil — APPROVED** [review]

**Pull request merged**
